### ISSUE TITLE
Change behavior of offsets

### ISSUE THREAD

**Issue by yamiteru** [yamiteru/isitfast#31]
It seems like V8 optimizes empty functions differently based on how hot they are.

We gather offsets once per every `Suite`. The number of iterations is determined by the `deviationPercent`.

The problem comes from for example offsets being stable after `2 500` iterations while the benchmark being stable after `500 000` iterations. The cost of the wrapper function seems to be much lower after `500 000` iterations than in `2 500` iterations.

What this causes is that if the benchmark take `2 ns` and the difference of wrapper function cost (`2 500` vs `500 000`) is `4 ns` then the result is `-2 ns` which is clamped to `0 ns` making it infinite.

---

One way to solve it is to run offsets AFTER EVERY benchmark and to run it N times based on how many times the benchmark was ran.

Another solution might be to dynamically change the koefficient based on the smallest number. If the koefficient is `1`, the offset `50 ns` and the results `[35, 64, 97, 122]` then the koefficient must be changed to less than `0.7` to make the results look like this `[1, 30, 63, 88]`. This way there is no infinite result and all the results still have the same difference between them (in percent). _The problem with this solution are benchmarks that should be infinite which is a benchmark of an empty function. Maybe in such a case the koefficient might be set to `0`._

**Comment by yamiteru** [comment]
So I've tried doing some tests regarding the number of iterations. Here are the results:

cycles | async cpu | async ram | sync cpu | sync ram
--- | --- | --- | --- | ---
50 | 77 | 448 | 51 | 256
2500 | 58 | 448 | 38 | 256
100000 | 22 | 304 | 24 | 256
10000000 | 20 | 304 | 21 | 256

It's very apparent that the more iterations the more the JS engine (in this case V8) optimizes even a simple empty function with literally nothing inside. Interestingly enough ram of async function changes somewhere between 2500 and 100000 iterations from 448 to 304.

This basically confirms my theory. So now the next step is to do iteration based offset calculations.

The easiest thing would be to run the offsets after all benchmarks have run, gather iterations and run offsets for each iteration amount and then just subtract the offset from benchmarks. But this would defeat the whole event-based and real-time thing.

So instead I'm gonna move offset calculations at the end of every benchmark. I'm gonna keep a map of iteration amounts and their offsets and check if the offsets already exist before actually running them. If all benchmarks take the same amount of iterations (which they usually do) it's still gonna calculate offsets just once. If however they differ then there will be an appropriate number of offsets based on iteration count.

Hopefully the JS engine won't optimize and remember the optimized version of one offset. Effectively leaking optimization from one offset calculation to the other one making them yet again unreliable. If I find out that's the case then I'll be forced to move the offset calculations to their own threads/JS realms.

**Comment by yamiteru** [comment]
So after moving the offsets after each benchmark as mentioned in the previous message I'm still struggling with the infinite results. Not as much as I did previously but it still sucks.

When it comes to slow/complex benchmarks that take thousands of nanoseconds it doesn't matter since the const of the benchmark wrapper function is gonna be small (depending on the number of iteration somewhere between 30 - 80 nanoseconds).

But when it comes to fast/small benchmarks where you care about every small increment it's very important for the bench lib to not introduce any noise in the form of the wrapper function perf cost. If the logic inside that function takes 4 nanoseconds and the wrapper itself takes 40 nanoseconds then all of the benchmarks will look basically the same since most of the nanoseconds don't come from them but from the wrapper function.

It might not be the best DX, but right now I think the only viable solution to this would be passing two functions down into the wrapper function. `start` and `end`. It'll move the actual benchmarking logic into the body of the wrapper function virtually getting rid of the cost (I'm not sure about RAM though).

This is how it could look like:

```ts
new Suite("Array loops")
  .add("for", function({ start, end }) {
    start();

    for(let i = 0; i < 10000; ++i) {
      tmp = i;
    }

    end();
  })
```  

Maybe the `start` and `end` function will have to be async requiring the wrapper function to also be async in order to await them. Also the functions themselves might also have some perf cost so it might not solve the problem at all.

The last option would be to introduce some sort of compilation step which would get rid of all the junk that is here for the DX and possible run every benchmark in it's own thread/JS realm but all that sounds way too complex. However what I'm after is the most statistically sounds JS/TS benchmarking lib so if there will be no other option then I guess I'll go the compilation route.

**Comment by yamiteru** [comment]
I've figured out that the way I was creating medians was flawed so I'm gonna redo it. Also even though Median Absolute Deviation is better for results it makes the benchmarks end too soon so in the end I'm gonna use Standard Deviation (but I'm gonna return both from the benchmark).

Also I've stumbled upon HDR Histograms. I don't think they can improve the way we run the benchmarks but it might be a nice addition to the data I return from each benchmark so I'm gonna add it.

**Comment by yamiteru** [comment]
This is how a benchmark of empty functions looks like now.

![image](https://user-images.githubusercontent.com/34185465/231880242-e6713f2e-d163-48d3-90a3-cd097282df60.png)

Previously the number of iterations were exactly the same for both. Now you can see that with Standard Deviation as opposed to Median Absolute Deviation it takes much longer for async to stabilize than sync. That's exactly what we want.

**Comment by yamiteru** [comment]
I've tried "compiling" benchmarks (using `Function`) in order to get rid of overhead of calling `now` and `memory` functions. It did work (but not as well as I've hoped for) but I couldn't execute benchmarks that used anything outside their scope (not very usable for testing libraries) so I reverted it back.

There is a way of "compiling" and executing code outside of the scope of the benchmark functions just like `benchmark.js` does it (save benchmarks into an object and access that object from inside the compiled function). I might implement this later once everything else is working and stable since like I've mentioned the difference was negligible.

At first the offsets were being computed before benchmarks were executed. Now I've tried computing them together with benchmarks. It seems that the best solution is gonna be computing them after all benchmarks have finished.

Let's say we have these tuples of benchmark and offset in nanoseconds: `[[36, 21], [56, 41]]`. After subtracting the offsets from benchmark results we get `[15, 15]`. Both are `15`. But that's wrong. Why is it wrong? Because the first one gets offset penalty of `21` and the second one gets `41` (given the algorithm that I use right now). Here we can argue that we can simply use the smaller offset `21` for both. But what if the order was different?

So the only reasonable thing to do is to run all benchmarks, figure out the smallest median and keep running the offset function until its median becomes smaller than the smallest benchmark result and only after that apply it to all of the benchmarks.

This will require the terminal to re-render after offsets are gathered and applied so I think I'll finally switch from my very underground `process.stdout` to something like [ink](https://github.com/vadimdemedes/ink).

_(I hope this is gonna be it because I've rewritten the whole library like 6 times already in the last week)_
